javascript/react/common-mistakes-in-react/dont-mutate-state.md

```diff
@@ -0,0 +1,81 @@
+---
+author: kapnobatai136
+
+levels:
+
+  - basic
+  
+  - medium
+  
+aspects:
+
+  - workout
+
+  - deep
+
+type: normal
+
+category: tip
+
+---
+
+# Don't mutate state
+
+---
+## Content
+
+Now that you know how to handle yourself with React, it is time to learn about some of the most common mistakes and how to avoid them.
+
+One of the most common mistakes one can make when using React is trying to directly modify the `state`. In a previous workout we have discussed how the `state` is considered immutable, and some methods of updating arrays or objects were presented. Consider the following example:
+
+```js
+const updateFeaturesList = (e, idx) => {
+  listFeatures[idx].checked =
+    e.target.checked;
+  setListFeatures(listFeatures);
+};
+```
+
+Although this code might seem correct at a first glance, the UI won't reflect this because the `state` is updated with the same object reference which doesn't trigger a re-render. Because the `state` has an asynchronous nature, mutating it might result in later `state` updates overriding the changes you have made directly. When using functional components, use the setter method returned by `useState()`:
```

Comment:
```suggestion
Although this code might seem correct at a first glance, the UI won't actually reflect this change because the `state` is updated on the same object reference, which doesn't trigger a re-render. Because the `state` updates can happen asynchronously (React might do this outside of our control), mutating it might result in later `state` updates overriding the changes you have made directly. To circumvent this when using functional components, use the setter method returned by `useState()`:
```